Fix ruby backwards compatibility

rafamanzo opened the pull request:
Prezento's build https://travis-ci.org/mezuro/prezento/builds/129486939 revealed this flaw.

We add those versions to travis' build matrix in order to avoid this in the future.

rafamanzo committed:
Fix availability check on older Rubies

The call to `system` we were using with Ruby 2.3 gives unexpected
behaviour under rubies 2.1.5 and 2.0.

Signed-off-by: Eduardo Araújo <duduktamg@hotmail.com>

rafamanzo committed:
Build against older rubies

This ensures backwards compatibility with key linux distros: Debian and
CentOS.

Signed-off-by: Eduardo Araújo <duduktamg@hotmail.com>

rafamanzo committed:
Remove unecessary cucumber ruby specific profiles

The default rake task implementation switches between profiles specific
for each Ruby version at runtime. But those profiles are not defined for
this project.

Accordingly to https://github.com/cucumber/cucumber-ruby/blob/b781bbf6b99d1d127e3468399a7e335f3cc99b5e/cucumber.yml
this just ignores jruby tagged scenarios which is not the case for this
project.

Thus those removing such switching has no losses to this project.

Commit:
Fix collector call to `system`

The call to `system` we were using with Ruby 2.3 gives unexpected
behaviour under rubies 2.1.5 and 2.0.

diegoamc commented on gem_tasks/cucumber.rake:
```diff
@@ -3,25 +3,8 @@
 require 'cucumber/rake/task'
 require 'cucumber/platform'
 
-class Cucumber::Rake::Task
```
Shouldn't this code removal be on another PR? I don't see how removing this code helps to fix a backwards compability problem.

rafamanzo replied:
Have a look at the commit message when you have the time: https://github.com/mezuro/kolekti_metricfu/pull/15/commits/b76bb041d4679e082b537a6976e7bde8bcd6a7cf

It helps by making possible to reproduce the problem at first.

diegoamc replied:
I've read the commit message before writing the comment. I've read it again and still don't see how this helps to reproduce the problem.

How did you reproduce it? By running Prezento's acceptance tests or by running kolekti_metricfu's acceptance tests? What happens if you run cucumber using an undefined profile?

do-you-dare commented:
@danielkza, it doesn't raise errors, but produces incorrect output under ruby 2.0.0 and 2.1.5. Using the more verbose one gives the right results.

Now, 2.0.0-p598. Running locally, it seems that the error stems from `reek`.

danielkza commented:
@dread-uo I figured out what the difference is. When passing `out` and `err` separately, Ruby has special behavior that knows to open each file read-write before passing it out to the child process. It seems before Ruby 2.2 (?) the same didn't apply when you passed `out` and `err` together, and `/dev/null` was opened read-only, and any write to stdout or stderr would cause weird stuff to happen.

strace of Ruby 2.1.5, first with separate parameters, then together:

```
open("/dev/null", O_WRONLY|O_CREAT|O_TRUNC, 0644) = 7
fstat(7, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
dup2(7, 1)                              = 1
fstat(1, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
close(7)                                = 0
open("/dev/null", O_WRONLY|O_CREAT|O_TRUNC, 0644) = 7
fstat(7, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
dup2(7, 2)                              = 2
fstat(2, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
close(7)                                = 0
execve("/bin/sh", ["sh", "-c", "echo asd | tee /dev/stderr"], [/* 14 vars */]) = 0
```

```
open("/dev/null", O_RDONLY)             = 7
fstat(7, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
dup2(7, 1)                              = 1
fstat(1, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
dup2(7, 2)                              = 2
fstat(2, {st_mode=S_IFCHR|0666, st_rdev=makedev(1, 3), ...}) = 0
close(7)                                = 0
execve("/bin/sh", ["sh", "-c", "echo asd | tee /dev/stderr"], [/* 14 vars */]) = 0
```

I'd rewrite the commit message to reflect this instead of saying it is a syntax error.

Commit:
Run metric_fu without reek

reek fails with some projects under ruby 2.0. As we don't use its
output, we'll simply run metric_fu without it.
